### PR TITLE
Enabling starting connection without protocol negotiation

### PR DESCRIPTION
**Pull request by moozzyk**
#932

### CODE REVIEW

**Comment by moozzyk** [comment]
This has questionable behavior in case of a client/server configuration mismatch - in some cases the server will just ignore the Negotiation message but in other cases it will fail with weird errors.

**Comment by davidfowl** [comment]
I'm not a fan of the "skip negotiate flag on the client" (aka `WithNoProtocolNegotiation`). Picking a protocol on the server is enough to make it use that specific protocol without negotiation but it isn't enough on the client. 

We should clarify a few things:
- When a client is configured with a specific transport an a specific protocol, why does it need to negotiation (is it because of the text vs binary decision?).
- When the server specifies a specific protocol, what sort of error does the client get if the client doesnt' send a negotiate and connects with the wrong protocol?

PS: It should be WithoutProtocolNegotiation 😄 .

**Comment by moozzyk** [comment]
> I'm not a fan of the "skip negotiate flag on the client"

We are on the same page. I initially tried adding a `skipNegotiate` parameter to `WithHubProtocol()` that the user could set but it is basically the same things only the number of overloads exploded. 
The main reason to have this flag is that if the server is not configured to use one protocol you have to send negotiate always to the server otherwise it will not know what protocol to use to handle your connection. 

If the server receives a wrong protocol it will not be able to parse the message and will throw `FormatException` which results in closing the connection. 

> It should be WithoutProtocolNegotiation

A very good point.

**Comment by moozzyk** [comment]
So, actually when the server is using MsgPack and sends JSon the server will very likely not close the connection. It might read the first x bytes as a VarInt and then if the message is too short it will throw the contents. This needs more thought because we don't seem to have enough information on the server side to tell the protocols apart just by message contents (and this does not seem to be correct as well).

**Comment by moozzyk** [comment]
Maybe could pass missing information in the transport-specific way. With websockets we could send it as a subprotocol, for SSE and longpolling we could use a header? HubConnection would set a feature on IConnection which the client transport would read and act accordingly. On the server side the server would read the value and set the feature when creating a connection. This feature could be then consumed by HubEndPoint....

**Comment by davidfowl** [comment]
Or maybe this feature is just a bust. We can revisit later.

File: src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionBuilderExtensions.cs [hunk 1]
```diff
@@ -72,5 +78,10 @@ public static IHubProtocol GetHubProtocol(this IHubConnectionBuilder hubConnecti
             hubConnectionBuilder.TryGetSetting<IHubProtocol>(HubConnectionBuilderDefaults.HubProtocolKey, out var hubProtocol);
             return hubProtocol;
         }
+
+        public static bool GetNoProtocolNegotiation(this IHubConnectionBuilder hubConnectionBuilder)
```
Comment (mikaelm12) [inline src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnectionBuilderExtensions.cs]: I think this reads better as `GetWithoutProtocolNegotiation`

**Comment by moozzyk** [comment]
This is a pit of failure. It is very easy to misconfigure things which may result in a range of issues - from abruptly closing the connection to connections that appear to be connected but throwing all the messages on the server side.

**Pull request closed without merging**